File: src/main/java/infrastructure/network/finnhub/FinnhubClient.java
package infrastructure.network.finnhub;

import com.google.gson.Gson;
import com.google.gson.GsonBuilder;
import domain.stock.Quote;
import infrastructure.network.finnhub.dto.FinnhubQuoteDto;

import java.io.IOException;
import java.net.URI;
import java.net.URLEncoder;
import java.net.http.HttpClient;
import java.net.http.HttpRequest;
import java.net.http.HttpResponse;
import java.nio.charset.StandardCharsets;

//===========================================================//
/**
 * Client class for communicating with the Finnhub API.
 * <p>
 *    This class is responsible for HTTP communication with Finnhub.
 *    It sends requests, receives JSON response, converts the response
 *    DTO into Quote domain model
 * </p>
 */
//===========================================================//

public final class FinnhubClient {
   /*===================================================*/
   /*===================================================*/
   // Private Field(s)

   private static final Gson s_GSON = new GsonBuilder()
                                        .setPrettyPrinting()
                                        .create();

   private final FinnhubConfig m_config;
   private final HttpClient m_httpClient;

   /*===================================================*/
   /*===================================================*/
   // Public Method(es)

   /**
    * Gets the latest quote for a given stock symbol.
    *
    * @param symbol stock ticker symbol, for example "NET", "AAPL", "MSFT"
    * @return latest quote as a clean domain object
    * @throws IOException if the Finnhub request fails
    * @throws InterruptedException if the HTTP request is interrupted
    */
   public Quote getQuote(final String symbol) throws IOException, InterruptedException {
      if(symbol == null || symbol.isEmpty()) {
         throw new IllegalArgumentException("Symbol is missing");
      }

      final String encodedSymbol = URLEncoder.encode(symbol, StandardCharsets.UTF_8);
      final String url = m_config.getBaseUrl() + "/quote?symbol=" + encodedSymbol;

      final HttpRequest request = HttpRequest.newBuilder()
                                    .uri(URI.create(url))
                                    .timeout(m_config.getTimeout())
                                    .header("X-Finnhub-Token", m_config.getApiKey())
                                    .GET()
                                    .build();

      final HttpResponse<String> response = m_httpClient.send(
        request,
        HttpResponse.BodyHandlers.ofString()
      );

      if(response.statusCode() < 200 || response.statusCode() >= 300){
         throw new IOException(
           "Finnhub request failed. Status: " + response.statusCode()
           + " Body: " + response.body()
         );
      }

      final FinnhubQuoteDto dto = s_GSON.fromJson(
        response.body(),
        FinnhubQuoteDto.class
      );

      if(dto == null){
         throw new IOException("Finnhub returned an empty quote response");
      }

      return dto.toDomain(symbol);
   }

   /*===================================================*/
   /*===================================================*/
   // Constructor(s)

   /**
    * Initializes Finnhub client with config
    *
    * @param config Finnhub configuration API KEY, base URL and timeout
    */
   public FinnhubClient(final FinnhubConfig config){
      if (config == null) {
         throw new IllegalArgumentException("Config is missing");
      }
      this.m_config = config;
      this.m_httpClient = HttpClient.newHttpClient();
   }
}

File: src/main/kotlin/infrastructure/network/finnhub/FinnhubClient.kt
package infrastructure.network.finnhub

import com.google.gson.Gson
import com.google.gson.GsonBuilder
import com.google.gson.JsonSyntaxException
import com.sun.tools.javac.tree.DCTree.isBlank
import domain.assets.security.SecurityIdentifier
import infrastructure.network.finnhub.dto.FinnhubQuoteDto
import infrastructure.network.finnhub.dto.FinnhubSymbolDto
import infrastructure.network.httpGetRequestBuilder
import java.io.IOException
import java.net.URLEncoder
import java.net.http.HttpClient
import java.net.http.HttpResponse
import java.nio.charset.StandardCharsets

//===========================================================//
/**
 * Client class for communicating with the Finnhub API.
 * 
 * This class is responsible for HTTP communication with Finnhub.
 * It sends requests, receives JSON response, converts the response
 * DTO into Quote domain model
 */
//===========================================================//

class FinnhubClient (
    private val m_Config: FinnhubConfig,
    private val m_HttpClient: HttpClient = HttpClient.newHttpClient(),
)
{
    companion object {
        private val s_GSON: Gson = GsonBuilder()
            .setPrettyPrinting()
            .create()
        private val s_HEADER_NAME = "X-Finnhub-Token"
        private const val s_HTTP_MIN = 200
        private const val s_HTTP_MAX = 299
    }

    /*===================================================*/
    /*===================================================*/
    // Public Method(es)

    /**
     * Gets the latest quote for a given stock symbol.
     *
     * @param identifier The security identifier containing the isin number
     * @return [Result] containing finnhub quote dto.
     */

    fun getQuoteAsync(identifier: SecurityIdentifier): Result<FinnhubQuoteDto> {
        val symbol = getSymbol(identifier.isin).getOrElse {
            error -> return Result.failure(error)
        }

        val encodedSymbol = URLEncoder.encode(symbol, StandardCharsets.UTF_8)

        return getJson("/quote?symbol=$encodedSymbol")
    }

    /*===================================================*/
    /*===================================================*/
    // Private Method(es)

    /**
     * Gets Ticker Symbol for a given isin number.
     *
     * @param isin International Securities Identification Number
     * @returns result ticker symbol
     */
    private fun getSymbol(isin: String): Result<String> {
        val encodedIsin = URLEncoder.encode(isin, StandardCharsets.UTF_8)

        val dto = getJson<FinnhubSymbolDto>("/search/?q=$encodedIsin").getOrElse{
            error -> return Result.failure(error)
        }

        val symbol = dto.result.firstOrNull()?.symbol

        if(symbol == null || symbol.isBlank()){
            return Result.failure(
                IllegalStateException("Finnhub did not return a ticker symbol")
            )
        }

        return Result.success(symbol)
    }

    /**
     * Sends Get request to the given Finnhub endpoint and parses JSON response
     *
     * @param endpoint finnhub endpoint, including query parameter
     * @return [Result] containing the dto
     */
    private inline fun <reified T> getJson(endpoint: String): Result<T> {
        val url = m_Config.baseUrl + endpoint

        val request = httpGetRequestBuilder(url,
            s_HEADER_NAME,
            m_Config.timeout,
            m_Config.apiKey
        )

        return try {
            val response: HttpResponse<String> = m_HttpClient.send(
                request,
                HttpResponse.BodyHandlers.ofString()
            )

            if (response.statusCode() !in s_HTTP_MIN..s_HTTP_MAX) {
                return Result.failure(
                    IOException("Finnhub request failed. Status code: ${response.statusCode()}, URL: $url")
                )
            }

            val body = response.body()

            if (body.isNullOrBlank()) {
                return Result.failure(
                    IOException("Finnhub returned an empty response. URL: $url")
                )
            }

            val dto = s_GSON.fromJson(body, T::class.java)

            if (dto == null) {
                Result.failure(IOException("Failed to parse Finnhub response. URL: $url"))
            } else {
                Result.success(dto)
            }

        } catch (exception: IOException) {
            Result.failure(exception)
        } catch (exception: JsonSyntaxException) {
            Result.failure(IOException("Finnhub returned invalid JSON. URL: $url", exception))
        } catch (exception: IllegalArgumentException) {
            Result.failure(IOException("Invalid Finnhub request URL: $url", exception))
        }
    }
}
